source/applications/initializers.md

```diff
@@ -0,0 +1,100 @@
+Initializers provide an opportunity to configure your application as it boots.
+
+There are two types of initializers: application initializers and application
+instance initializers.
+
+Application initializers are run as your application boots, and provide the
+primary means to configure [dependency injections](./dependency-injection) in
+your application.
+
+Application instance initializers are run as an application instance is loaded.
+They provide a way to configure the initial state of your application, as well
+as to set up dependency injections that are local to the application instance
+(e.g. A/B testing confurations).
+
+Initializers can perform any configuration operations, as long as they are
+synchronous. The user experience is degraded by attempting to block the booting
```

Comment:
> I thoughts initializers could defer? 

Application initializers (not application instance initializers) could technically call `deferReadiness`, perform something async, and then call `advanceReadiness`.

However, with the application / instance split (and the accompanied registry / container split), it is no longer possible to do anything state related because there isn't even a container that's accessible from an application initializer.

I think the use cases for advancing / deferring readiness are so slim now that it's almost always an anti-pattern that's better handled in a route hook (see the comment below). However, I'm open to hearing objections :)

> For example, if you wanted to lazily load some code on application boot then register it you could use an initializer.

As for lazy code loading, this should eventually be handled by engines, but the asynchrony will be integrated into the router. I'd suggest that in the future lazy code loading shouldn't defer loading of an application, but will be reason to create a lazy-loaded engine. Obviously, we don't want to confuse the docs with engines at this point - just clarifying my position.